x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceValueSource.java

```diff
@@ -0,0 +1,137 @@
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSource.Bytes;
+import org.elasticsearch.search.aggregations.support.ValuesSource.Numeric;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Interface to extract values from Lucene in order to feed it into the MapReducer.
+ */
+public abstract class MapReduceValueSource {
+
+    @FunctionalInterface
+    public interface ValueSourceSupplier {
+        MapReduceValueSource build(ValuesSourceConfig config, int id);
+    }
+
+    public static class Field implements Writeable {
+        private final String name;
+        private final int id;
+
+        public Field(String name, int id) {
+            this.name = name;
+            this.id = id;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public int getId() {
+            return id;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(name);
+            out.writeInt(id);
+
+        }
+    };
+
+    private final Field field;
+    private final DocValueFormat format;
+
+    abstract Tuple<Field, List<Object>> collect(LeafReaderContext ctx, int doc) throws IOException;
+
+    MapReduceValueSource(ValuesSourceConfig config, int id) {
+        String fieldName = config.fieldContext() != null ? config.fieldContext().field() : null;
+
+        if (Strings.isNullOrEmpty(fieldName)) {
+            throw new IllegalArgumentException("scripts are not supported");
+        }
+
+        this.field = new Field(fieldName, id);
+        this.format = config.format();
+    }
+
+    public Field getField() {
+        return field;
+    }
+
+    DocValueFormat getFormat() {
+        return format;
+    }
+
+    public static class KeywordValueSource extends MapReduceValueSource {
+        private final ValuesSource.Bytes source;
+
+        public KeywordValueSource(ValuesSourceConfig config, int id) {
+            super(config, id);
+            this.source = (Bytes) config.getValuesSource();
+        }
+
+        @Override
+        public Tuple<Field, List<Object>> collect(LeafReaderContext ctx, int doc) throws IOException {
+            SortedBinaryDocValues values = source.bytesValues(ctx);
+
+            if (values.advanceExact(doc)) {
+                int valuesCount = values.docValueCount();
+                List<Object> objects = new ArrayList<>(valuesCount);
+
+                for (int i = 0; i < valuesCount; ++i) {
+                    objects.add(getFormat().format(values.nextValue()));
```

Comment:
I am very concerned about doing formatting here.  Collecting aggregations is one of the hottest code paths we have.  Formatting especially is not at all speed optimized, and can be quite costly.  Furthermore, formatting at collect time will cause a massive repetition of work, as repeated values are formatted each time they are read, and many of those values might not be used in the final output.

Reply:
good point, this is a leftover. As discussed last week I _removed_ support for `format`, therefore this is dead code.

Reply:
formatting is now applied on the result side. The implementation is similar to multi terms. I still don't see a need to bring back the `format` parameter, but nevertheless formatting is required to properly decode `BytesRefs` which might be differently encoded.

(I will add some more unit tests tomorrow, just did not want to hold off this change)